src/AppBundle/EventSubscriber/RateLimitSubscriber.php

```diff
@@ -0,0 +1,150 @@
+            return;
+        }
+
+        $request = $event->getRequest();
+
+        $this->checkBlacklist($request);
+
+        // Build and fetch cache key based on session ID and requested URI,
+        //   removing any reserved characters from URI.
+        $uri = preg_replace('/[^a-zA-Z0-9,\.]/', '', $request->getRequestUri());
+        $sessionId = $request->getSession()->getId();
+        $cacheKey = 'ratelimit.'.$sessionId.'.'.$uri;
+        $cache = $this->container->get('cache.app');
+        $cacheItem = $cache->getItem($cacheKey);
+
+        // If increment value already in cache, or start with 1.
+        $count = $cacheItem->isHit() ? (int) $cacheItem->get() + 1 : 1;
+
+        // Check if limit has been exceeded, and if so, throw an error.
+        if ($count > $this->rateLimit) {
+            $this->denyAccess($request, 'Exceeded rate limitation');
+        }
+
+        // Reset the clock on every request.
+        $cacheItem->set($count)
+            ->expiresAfter(new DateInterval('PT'.$this->rateDuration.'M'));
+        $cache->save($cacheItem);
+    }
+
+    /**
+     * Check the request against blacklisted URIs and user agents
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     */
+    private function checkBlacklist(\Symfony\Component\HttpFoundation\Request $request)
+    {
+        // First check user agent and URI blacklists
+        if ($this->container->hasParameter('request_blacklist')) {
+            $blacklist = $this->container->getParameter('request_blacklist');
+            // User agents
+            if (is_array($blacklist['user_agent'])) {
+                foreach ($blacklist['user_agent'] as $ua) {
+                    if (strpos($request->headers->get('User-Agent'), $ua) !== false) {
+                        $this->denyAccess($request, "Matched blacklisted user agent `$ua`");
+                    }
+                }
+            }
+            // URIs
+            if (is_array($blacklist['uri'])) {
+                foreach ($blacklist['uri'] as $uri) {
+                    if (strpos($request->getRequestUri(), $uri) !== false) {
+                        $this->denyAccess($request, "Matched blacklisted URI `$uri`");
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Throw exception for denied access due to spider crawl or hitting usage limits.
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @param string $logComment Comment to include with the log entry.
+     * @todo i18n
+     */
+    private function denyAccess(\Symfony\Component\HttpFoundation\Request $request, $logComment = '')
+    {
+        // Log the denied request
+        $logger = $this->container->get('monolog.logger.rate_limit');
+        $logger->info(
+            "<URI>: " . $request->getRequestUri() .
+            ($logComment != '' ? "\t<Reason>: $logComment" : '') .
+            "\t<User agent>: " . $request->headers->get('User-Agent')
+        );
```

Comment:
This doesn't come out looking so pretty, but apparently it requires a lot of confusing configuration to get Monolog to allow multi-line log comments. It's still legible which is the important part. See https://stackoverflow.com/questions/29312746/symfony2-how-to-log-multiline-entries-with-monolog

We might also want to add log rotation.